options/ansi/generic/stdio-stubs.cpp

```diff
@@ -1003,46 +1003,69 @@ void perror(const char *string) {
 	fprintf(stderr, "%s\n", strerror(error));
 }
 
-// POSIX unlocked I/O extensions.
+// POSIX extensions.
 
-// GLIBC extensions.
+ssize_t getline(char **line, size_t *n, FILE *stream) {
+	return getdelim(line, n, '\n', stream);
+}
 
-ssize_t getline(char **line, size_t *n, FILE *file_base) {
+ssize_t getdelim(char **line, size_t *n, int delim, FILE *stream) {
 	// Otherwise, we cannot store the buffer / size.
 	if(!line || !n) {
 		errno = EINVAL;
 		return -1;
 	}
 
-	char *buffer = nullptr;
-	size_t capacity = 0;
-	if(*line) {
-		buffer = *line;
-		capacity = *n;
-	}
+	constexpr size_t chunk = 1024;
 
-	if(!capacity) {
-		buffer = reinterpret_cast<char *>(getAllocator().allocate(1024));
-		capacity = 1024;
-	}
+	char *buffer = *line;
+	size_t capacity = *n, nwritten = 0;
 
-	if(!fgets(buffer, capacity, file_base))
+	auto file = static_cast<mlibc::abstract_file *>(stream);
+	frg::unique_lock<FutexLock> lock(file->_lock);
+
+	// Avoid allocating if we've already hit the end
+	auto c = fgetc_unlocked(stream);
+	if (c == EOF || ferror(stream)) {
 		return -1;
+	} else {
+		file->unget(c);
+	}
+
+	while (true) {
+		// Fill the buffer
+		while (capacity > 0 && nwritten < capacity - 1) {
```

Comment:
What happens when capacity is not 0, but the buffer is still NULL? Maybe that wouldn't happen but might still be a good case to catch.